dynamic_instrumentation/src/main.rs

```diff
@@ -0,0 +1,413 @@
+    let sysroot: &Utf8Path = sysroot.try_into()?;
+    at_args.extend(["--sysroot".into(), sysroot.as_str().into()]);
+    RunCompiler::new(&at_args, &mut MirTransformCallbacks)
+        .run()
+        .map_err(|_| eyre!("`rustc` failed"))?;
+    INSTRUMENTER.finalize(metadata).map_err(|e| eyre!(e))?;
+    Ok(())
+}
+
+/// Delete all files that we want to be regenerated
+/// so that we always have fully-up-to-date, non-incremental metadata.
+fn delete_metadata_and_dependencies(
+    info: &InstrumentInfo,
+    cargo_metadata: &Metadata,
+) -> eyre::Result<()> {
+    if let Err(e) = fs_err::remove_file(&info.metadata) {
+        if e.kind() != ErrorKind::NotFound {
+            return Err(e.into());
+        }
+    }
+
+    // TODO(kkysen) Figure out a way to know which profile is being used
+    // Until then, just search them all and delete all of them.
+
+    // TODO(kkysen) We probably have to delete binaries that have different names from the crates.
+
+    // Delete all executables in `target/${profile}/deps/` starting with the crate name + `-`.
+    let target_read_dir = match cargo_metadata.target_directory.read_dir() {
+        Ok(read_dir) => Ok(read_dir),
+        Err(e) => if e.kind() == ErrorKind::NotFound {
+            return Ok(());
+        } else {
+            Err(e)
+        }
+    }?;
+    for profile_dir in target_read_dir {
+        let profile_dir = profile_dir?;
+        if !profile_dir.file_type()?.is_dir() {
+            continue;
+        }
+        for artifact in profile_dir.path().join("deps").read_dir()? {
+            let artifact = artifact?;
+            if !artifact.file_type()?.is_file() {
+                continue;
+            }
+            let file_name = artifact.file_name();
+            let file_name = file_name.to_str();
+            for crate_target in &info.crate_targets {
+                let prefix = format!("{}-", crate_target.crate_name.as_str());
+                // [`Path::starts_with`] only checks whole components at once,
+                // and `OsStr::starts_with` doesn't exist yet.
+                if file_name
+                    .map(|name| name.starts_with(&prefix))
+                    .unwrap_or_default()
+                {
+                    let artifact = artifact.path();
+                    if artifact.is_executable() {
+                        fs_err::remove_file(&artifact)?;
+                    }
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn main() -> eyre::Result<()> {
+    color_eyre::install()?;
+    env_logger::init();
+
+    let rustc_wrapper_var = "RUSTC_WRAPPER";
+    let instrument_info_var = "C2RUST_INSTRUMENT_INFO";
+
+    let own_exe = env::current_exe()?;
+
+    let wrapping_rustc = env::var_os(rustc_wrapper_var).as_deref() == Some(own_exe.as_os_str());
+    if wrapping_rustc {
+        let sysroot = get_sysroot()?;
+        let at_args = env::args().skip(1).collect::<Vec<_>>();
```

Comment:
Cargo passes rustc as the first arg. Is that the full path to rustc? If so, could we use that to get the sysroot and not rely on rustup at all?

Reply:
Unfortunately, it's not, or at least, it's not usually (it probably can be).  It's normally just `rustc`, and `$RUSTC` is only set to `rustc`, too (I think there was a change to this in the latest `rustup` ([`CHANGELOG.md`](https://github.com/rust-lang/rustup/blob/master/CHANGELOG.md#1251---2022-07-12)) where the full path was specified in `$RUSTC` as an optimization, but it got reverted due to some people depended on it).

Reply:
Note that we can look up the sysroot without explicitly depending on `rustup` with `rustc --print sysroot` (I haven't implemented that part yet), the current way just uses a `rustup`-dependent fast-path.

Reply:
`get_sysroot_slow` is now implemented in https://github.com/immunant/c2rust/pull/554/commits/628266b0ca0cb7347ff981d6e806d36db41f347c.